Add some tests that exercise pytype to travis

alvarocaceres opened the pull request.

gvanrossum commented:
I like this idea, but it does slow down the tests quite a bit. (And will do so even more once you add pytype support for Python 3 -- they claim to support it.)

Also, isn't there a Travis feature "matrix" that can simplify the syntax used here?

Finally, I don't really like that you had to change the #! line in runtests.py to use python instead of python3.

Maybe it's better to refactor runtests.py into a library module that discovers all the relevant files/modules and separate main programs, one to run mypy and one to run pytype?

matthiaskramm commented:
I also noticed that the pytd tool will default to optimizing the pyi it reads. We can turn that off, that should make it run faster.

alvarocaceres commented:
Looked at keeping python3 instead of python, but didn't find a way to get travis to provide both interpreters in the same environment, and pytype needs a py2 interpreter to run.

Didn't know about the matrix stuff, using it now so the yaml file should be more readable

Added some hacky parallelization to speed up pytype tests

I'm happy to split runtests into separate files, what kind of directory structure do you want?

matthiaskramm commented:
It looks like the pytype check is faster than the mypy check now (0:42 vs 1:16).

alvarocaceres commented:
yep, turning off optimizations in pytd made a big difference. I think the remaining issues from Guido are

- I switched runtests to be a py2and3 program instead of py3 (wonky syntax for prints, etc)
- pytype and mypy tests are in a single file, might be nicer to split them

Guido, are these still sticking points or is the current state good enough? If still an issue, I can make a tests/ directory and split things into a couple of files, so the mypy file can be in py3 syntax. Would still keep a common main().

gvanrossum commented:
I still think you should just have two different programs, one to run mypy tests and one to run pytype tests. (And if travis expects a single toplevel thing, make that a tiny shell script.)

Also, the list of pytype exceptions could be better off as a separate text file.

alvarocaceres commented:
Ok, I've split things into separate files, PTAL

gvanrossum commented on .travis.yml:
```diff
@@ -1,15 +1,18 @@
+matrix:
+  include:
+    - python: "3.5"
+      env: TEST_CMD="./tests/mypy_test.py -p3.5 -p3.4 -p3.3 -p3.2 -p2.7"
```
You don't need the -p flags here any more -- what you have is the default set.

gvanrossum commented on tests/pytype_blacklist.txt:
```diff
@@ -0,0 +1,59 @@
+# Pytype blacklist. Files will not be tested with pytype.
```
@matthiaskramm Why are so many files blacklisted? Is there any incentive to reduce this list?

gvanrossum commented on tests/mypy_test.py:
```diff
@@ -22,7 +22,7 @@
 parser.add_argument('-n', '--dry-run', action='store_true', help="Don't actually run mypy")
 parser.add_argument('-x', '--exclude', type=str, nargs='*', help="Exclude pattern")
-parser.add_argument('-p', '--python-version', type=str, nargs='*',
+parser.add_argument('-p', '--python-version', type=str, action='append',
```
Just curious -- why did this need to change? I'm unsure about the difference, or why -x and filter don't need to change.

alvarocaceres replied:
no good reason, I misunderstood how argparse works, submitted #358 to undo

gvanrossum commented:
Thanks! Apart from the above nit (and two questions) this LGTM.

matthiaskramm commented:
> @matthiaskramm Why are so many files blacklisted? Is there any incentive to reduce this list?

We're still working on our `pyi` parser. It evolved from a `pytd` parser, so it's a bit more strict in the syntax it accepts, and doesn't yet cope too well with the syntactical richness of typeshed.

We're working on improving it, which should shrink, and hopefully even eliminate, the blacklist.
